### PR TITLE
Fix #10103 - CheckConvexity removes colinear vertices but stops when surface is determined non-convex, causing vertex size mismatch fatal error

### LINKED ISSUE #10103: CheckConvexity removes colinear vertices but stops when surface is determined non-convex, causing vertex size mismatch fatal error
Issue overview
--------------

Issue was reported on Unmethours. Ceiling and floors are perfectly matching with 14 vertices, yet a fatal error is issued.

This is the floor:

image

And logically, the reversed ceiling:

image

CheckConvexity actually does not only check for convexity, it also removes colinear vertices. It processes vertices in the order shown above. For the floor, it detects all the "lower" (in surface coordinates) vertices as being colinear before it gets to the test on points (10, 11, 12), detects a non-convex surface and bails out via  a `break`. It therefore makes this one 7 vertices (really only 6 are needed...)

When it processes the ceiling, it goes to check points (3, 4, 5) and detects a non-convex surface and breaks. Therefore all of the colinear points have not been processed yet and it ends up with 14 vertices (untouched).

```
   ** Warning ** CheckConvexity: Surface="FACE 348", vertex 2 is colinear with previous and next.
   ** Warning ** CheckConvexity: Surface="FACE 348", vertex 3 is colinear with previous and next.
   ** Warning ** CheckConvexity: Surface="FACE 348", vertex 4 is colinear with previous and next.
   ** Warning ** CheckConvexity: Surface="FACE 348", vertex 5 is colinear with previous and next.
   ** Warning ** CheckConvexity: Surface="FACE 348", vertex 6 is colinear with previous and next.
   ** Warning ** CheckConvexity: Surface="FACE 348", vertex 7 is colinear with previous and next.
   ** Warning ** CheckConvexity: Surface="FACE 348", vertex 8 is colinear with previous and next.
   ** Warning ** CheckConvexity: Surface="FACE 348" has [7] collinear points that have been removed.
   ** Warning ** CheckConvexity: Surface="FACE 348": The vertex points has been reprocessed as Sides = 7
   ** Severe  ** RoofCeiling:Detailed="FACE 8", Vertex size mismatch between base surface :FACE 8 and outside boundary surface: FACE 348
   **   ~~~   ** The vertex sizes are 14 for base surface and 7 for outside boundary surface. Please check inputs.
```

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): All
 - Version of EnergyPlus (if using an intermediate build, include SHA): 23.1.0, develop 07b51a2c1ca55df2ec5ff297ef9514aa5ce4236a
 - Unmethours : https://unmethours.com/question/93600/the-vertex-sizes-are-7-for-the-base-surface-and-14-for-the-outside-boundary-surface-please-check-inputs/

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #10103

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 1]
```diff
@@ -15788,30 +15788,33 @@ namespace SurfaceGeometry {
...
-                        ShowContinueError(state, format("...vertex {}=[{:.2R},{:.2R},{:.2R}]", Np2, X(n + 2), Y(n + 2), Z(n + 2)));
-                        // ShowContinueError(state, format("...theta angle=[{:.6R}]", Theta));
-                        // ShowContinueError(state, format("...last theta angle=[{:.6R}]", LastTheta));
-                    }
+                if (state.dataGlobal->DisplayExtraWarnings && surfaceTmp.ExtSolar &&
+                    (state.dataHeatBal->SolarDistribution != DataHeatBalance::Shadowing::Minimal) &&
+                    // Warn only once
+                    surfaceTmp.IsConvex) {
+                    ShowWarningError(state,
+                                     format("CheckConvexity: Zone=\"{}\", Surface=\"{}\" is non-convex.",
```
Comment (jmarrec) [inline src/EnergyPlus/SurfaceGeometry.cc:15796]: When warning, do it only once (when surfaceTmp.IsConvex, before that's set to false)

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 2]
```diff
@@ -15788,30 +15788,33 @@ namespace SurfaceGeometry {
...
+                    ShowContinueError(state, format("...vertex {}=[{:.2R},{:.2R},{:.2R}]", Np1, X(n + 1), Y(n + 1), Z(n + 1)));
+                    ShowContinueError(state, format("...vertex {}=[{:.2R},{:.2R},{:.2R}]", Np2, X(n + 2), Y(n + 2), Z(n + 2)));
+                    // ShowContinueError(state, format("...theta angle=[{:.6R}]", Theta));
+                    // ShowContinueError(state, format("...last theta angle=[{:.6R}]", LastTheta));
                 }
                 surfaceTmp.IsConvex = false;
-                break;
+                // #10103 - We do not want to break early, because we do want to consistently remove colinear vertices
+                // to avoid potential vertex size mismatch fatal errors
+                // break;
```
Comment (jmarrec) [inline src/EnergyPlus/SurfaceGeometry.cc:15817]: And do NOT break, so we ensure we do drop all colinear vertices and do not end up with a vertex size mistmatch.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 3]
```diff
@@ -3274,6 +3274,86 @@ TEST_F(EnergyPlusFixture, SurfaceGeometry_CheckConvexityTest_ASHRAE901_Hospital_
...
+    //    │          │          │
+    //    │          │          │
+    //    │5         │          │
+    //    ├──────────┘          │
+    //    │           6         │
+    //    │                     │
+    //    │                     │
+    //    │                     │10
+    //    └────x─────x─────x────┴─────►
+    //    4     3      2      1           x
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:3295]: Unit test case. We expect 1, 2, 3, and 8 to be dropped, resulting in 6 vertices in the end.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 4]
```diff
@@ -3274,6 +3274,86 @@ TEST_F(EnergyPlusFixture, SurfaceGeometry_CheckConvexityTest_ASHRAE901_Hospital_
...
+
+        ceilingSurface.Vertex = ceilingVertices;
+
+        CheckConvexity(*state, ceilingSurfNum, ceilingSurface.Sides);
+
+        EXPECT_EQ(6, ceilingSurface.Sides);
+        EXPECT_FALSE(ceilingSurface.IsConvex);
+    }
+
+    EXPECT_EQ(floorSurface.Sides, ceilingSurface.Sides);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:3354]: Before fix:

```
[ RUN      ] EnergyPlusFixture.SurfaceGeometry_CheckConvexity_ColinearStability
/home/julien/Software/Others/EnergyPlus/tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:3331: Failure
Expected equality of these values:
  6
  floorSurface.Sides
    Which is: 8
/home/julien/Software/Others/EnergyPlus/tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:3350: Failure
Expected equality of these values:
  6
  ceilingSurface.Sides
    Which is: 9
/home/julien/Software/Others/EnergyPlus/tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:3354: Failure
Expected equality of these values:
  floorSurface.Sides
    Which is: 8
  ceilingSurface.Sides
    Which is: 9
```

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc (hunk 3 above)
Comment (jmarrec) [reply tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:3295]: Before the fix, the floor is checked in the order show above. so it drops 2 and 3, then finds a non convex situation when checking (5, 6, 7) and bails. We end up with 8 vertices.

For the ceiling, the vertices are reversed. So it drops 8 (shown above, 3 in actual order), finds the non convex situation at (7, 6, 5) (shown above, (4, 5, 6 in actual order)) and bails. We end up with 9 vertices.

(I explained the same on #10103 when I reported the defect)

Comment (Myoldmopar) [reply tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:3295]: Ahhh this makes sense and is a phenomenal demonstration of the problem and test of the routine.  Thanks @jmarrec !

@mjwitte this looks good, but feel free to take a look before this merges.

**Comment by Myoldmopar** [comment]
I don't think we need to hold this up any longer, this is great.  Thanks @jmarrec !

**Pull request merged**